In [583]:
import pandas as pd
import tsam
import plotly.express as px

from pathlib import Path

# Entry point
This notebook assumes data cleaning (normalizaion/scaling, aggregation) has been already applied to the datasets.

# Data path config

In [584]:
# The location of the current file
cur_location = Path().resolve()

# Path to data location from this file's path point of view
data_location = cur_location / "../data/"

# Data Loading

You need to match the separator the CSV files are using in **sep** argument, e.g. in the provided datasets, some use ',' delimiter, some use ';', this needs to be specified in the **sep** argument when reading data.

In [585]:
demand_EU = pd.read_csv(data_location / "Demand_ENTSO_E.csv", sep=";")
demand_EU.head()

,snapshot,AL,AT,BA,BE,BG,CH,CZ,DE,DK,...,NL,NO,PL,PT,RO,RS,SE,SI,SK,UK
0,01.01.2025 00:00,459,3871,609,6691,2374,4603,4029,28833,2606,...,9796,11525,11852,3591,3645,2578,9701,728,1823,18285
1,01.01.2025 01:00,428,3717,571,6327,2292,4711,3989,27781,2511,...,9535,11586,11284,3480,3532,2474,9551,699,1780,18154
2,01.01.2025 02:00,396,3522,539,6128,2237,4861,3885,26917,2426,...,9266,11540,10983,3289,3467,2316,9377,666,1733,16545
3,01.01.2025 03:00,376,3488,523,5982,2206,4810,3855,27016,2400,...,9013,11517,10827,3126,3438,2175,9417,648,1729,14918
4,01.01.2025 04:00,367,3598,526,5967,2194,4725,3865,26870,2405,...,9040,11602,10754,3007,3428,2059,9560,664,1724,14200


In [586]:
onwind_cf = pd.read_csv(data_location / "onwind_capacity_factors.csv", sep=",")
onwind_cf.head()

,snapshot,AL_onwind_2025,AT_onwind_2025,BA_onwind_2025,BE_onwind_2025,BG_onwind_2025,CH_onwind_2025,CZ_onwind_2025,DE_onwind_2025,DK_onwind_2025,...,NO_onwind_2025,PL_onwind_2025,PT_onwind_2025,RO_onwind_2025,RS_onwind_2025,SE_onwind_2025,SI_onwind_2025,SK_onwind_2025,UA_onwind_2025,XK_onwind_2025
0,01.01.2025 00:00,0.0,0.026443,0.0,0.968038,0.027777,0.030767,0.419450,0.716840,1.000000,...,0.131214,0.873780,0.116221,0.051165,0.0,0.297268,0.013709,0.114870,0.346401,0.0
1,01.01.2025 01:00,0.0,0.029090,0.0,0.975868,0.027310,0.030301,0.426459,0.728613,0.999992,...,0.109802,0.897004,0.109594,0.054343,0.0,0.286711,0.015315,0.114240,0.345689,0.0
2,01.01.2025 02:00,0.0,0.033465,0.0,0.983856,0.028493,0.029116,0.443365,0.743509,0.999972,...,0.092031,0.911984,0.105639,0.062091,0.0,0.278833,0.019344,0.110237,0.332158,0.0
3,01.01.2025 03:00,0.0,0.034560,0.0,0.988063,0.030633,0.029512,0.467903,0.760476,0.999996,...,0.078412,0.923292,0.107288,0.067657,0.0,0.271188,0.023585,0.110385,0.326958,0.0
4,01.01.2025 04:00,0.0,0.036894,0.0,0.987477,0.034158,0.031685,0.495288,0.771249,1.000000,...,0.071422,0.927705,0.103116,0.072680,0.0,0.259012,0.025262,0.112671,0.322901,0.0


In [587]:
hydro_inflow = pd.read_csv(
    data_location / "hydro_inflow_scaled_2025.csv", sep=";"
)
hydro_inflow.head()

,snapshot,AT_hydro_2025,BA_hydro_2025,BE_hydro_2025,BG_hydro_2025,CH_hydro_2025,CZ_hydro_2025,DE_hydro_2025,ES_hydro_2025,FI_hydro_2025,...,RO_hydro_2025.1,RS_hydro_2025,RS_hydro_2025.1,RS_hydro_2025.2,SI_hydro_2025,UA_hydro_2025,UK_hydro_2025,XK_hydro_2025,SE_hydro_2025,AL_hydro_2025
0,01.01.2025 00:00,230,83,2,49,411,39,54,922,420,...,56,2,2,8,135,556,43,3,4570,155
1,01.01.2025 01:00,229,83,2,49,411,40,54,922,420,...,56,2,2,8,135,555,43,3,4569,152
2,01.01.2025 02:00,229,83,2,49,412,41,54,922,421,...,56,2,2,8,135,555,43,3,4562,151
3,01.01.2025 03:00,228,83,2,49,412,41,54,922,422,...,56,2,2,8,135,555,43,3,4555,151
4,01.01.2025 04:00,228,83,2,49,412,41,54,922,423,...,56,2,2,8,135,555,43,3,4545,150


In [588]:
ror_cf = pd.read_csv(data_location / "ror_capacity_factors.csv", sep=",")
ror_cf.head()

,snapshot,AL_ror_2025,AT_ror_2025,BA_ror_2025,BE_ror_2025,BG_ror_2025,CH_ror_2025,CZ_ror_2025,DE_ror_2025,EE_ror_2025,...,NO_ror_2025,PL_ror_2025,PT_ror_2025,RO_ror_2025,RS_ror_2025,SE_ror_2025,SI_ror_2025,SK_ror_2025,UK_ror_2025,XK_ror_2025
0,01.01.2025 00:00,0.06198,0.304260,0.113561,0.687213,0.132907,0.266242,0.128353,0.596077,0.0,...,0.592028,0.503166,0.140826,0.371941,0.428571,0.35421,0.090094,0.232892,0.19516,0.07694
1,01.01.2025 01:00,0.06074,0.300330,0.149166,0.687989,0.132323,0.264119,0.406737,0.591670,0.0,...,0.585834,0.503452,0.062204,0.369386,0.411529,0.35409,0.087915,0.231347,0.19512,0.07694
2,01.01.2025 02:00,0.06031,0.300783,0.155915,0.626787,0.129821,0.262051,0.430766,0.584802,0.0,...,0.578365,0.502625,0.059455,0.363477,0.338847,0.35360,0.087664,0.233002,0.19510,0.07682
3,01.01.2025 03:00,0.06016,0.306833,0.111967,0.609455,0.130377,0.260064,0.438508,0.577704,0.0,...,0.576665,0.502492,0.044515,0.359085,0.249123,0.35307,0.087212,0.230574,0.19506,0.07675
4,01.01.2025 04:00,0.05997,0.305347,0.111861,0.587503,0.135743,0.258158,0.438407,0.575100,0.0,...,0.580272,0.502750,0.096728,0.359085,0.246115,0.35229,0.087112,0.230022,0.19502,0.07684


In [589]:
solar_cf = pd.read_csv(data_location / "solar_capacity_factors.csv", sep=";")
solar_cf.head()

,snapshot,AL_solar_2025,AT_solar_2025,BA_solar_2025,BE_solar_2025,BG_solar_2025,CH_solar_2025,CZ_solar_2025,DE_solar_2025,DK_solar_2025,...,NO_solar_2025,PL_solar_2025,PT_solar_2025,RO_solar_2025,RS_solar_2025,SE_solar_2025,SI_solar_2025,SK_solar_2025,UA_solar_2025,XK_solar_2025
0,01.01.2025 00:00,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1,01.01.2025 01:00,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2,01.01.2025 02:00,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,01.01.2025 03:00,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4,01.01.2025 04:00,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


Let's for the sake of example work with only Germany data here.

In [ ]:
# Slicing Germany data only
demand_DE = demand_EU[["snapshot", "DE"]]
solar_cf_DE = solar_cf[["snapshot", "DE_solar_2025"]]
onwind_cf_DE = onwind_cf[["snapshot", "DE_onwind_2025"]]
ror_cf_DE = ror_cf[["snapshot", "DE_ror_2025"]]
hydro_inflow_DE = hydro_inflow[["snapshot", "DE_hydro_2025"]]

# Name, data pairs for the data we will be working with
data = [
    ("demand_DE", demand_DE),
    ("solar_cf_DE", solar_cf_DE),
    ("onwind_cf_DE", onwind_cf_DE),
    ("ror_cf_DE", ror_cf_DE),
    ("hydro_inflow_DE", hydro_inflow_DE),
]

# Snapshot feature handling
Snapshots need to be **set as index** and **converted to datetime** data type, since this is what TSAM expects.

In [591]:
def set_new_index(df: pd.DataFrame, new_index_col: str = "snapshot"):
    """
    Takes a dataframe and sets the provided new_index_col as its index
    """

    # Setting index of the DF to the snapshot column
    df.set_index(df[f"{new_index_col}"], inplace=True)

    # Dropping snapshot column
    df.drop(columns=[f"{new_index_col}"], inplace=True)
    return df


# Setting the snapshot column as the new index
data = [(df_name, set_new_index(df, "snapshot")) for df_name, df in data]

print("Index columns were swapped successfully!")

Index columns were swapped successfully!


We also need to convert the indexes to datetime data type. In our case, snapshots (D/M/Y) start with the day, so we specify that in the dayfirst argument, which needs to understand the format, i.e. whether it's (D/M/Y) or (Y/M/D), etc.

In [592]:
# Converting index to datetime type
for df_name, df in data:
    df.index = pd.to_datetime(df.index, dayfirst=True)

print("Successfully converted indexes into datetime dtype!")

Successfully converted indexes into datetime dtype!


# Data Validation
We need to run some basic data validation so that TSAM does not silently break.

Let's make sure our data follows:
- Absence of duplicates
- Absence of NaNs
- Absence of non-numeric columns
- Full-year coverage
- Capacity-factor columns remain in the expected 0-1 range

## Duplicates, NaNs, numeric columns & full-year coverage
Let's write a validator function. It will take a dataset, name of the dataset, year of the data, and a period of the data snapshots.
The function will raise an error if any of the validations do not pass, if all validations are passed – print success.

In [593]:
def validate_df(df: pd.DataFrame, name: str, year_to_check: int = 2025, period: str = "h"):
    """
    Runs data validation, throws an error if at least 1 of these is true:
    found duplicates, NaNs, non-numeric columns or data does not cover all 8760 hours of the given year.
    """

    # Set expected range with provided period
    expected_index = pd.date_range(
        start=f"{year_to_check}-01-01 00:00",
        end=f"{year_to_check}-12-31 23:00",
        freq=period,
    )

    # Raise an error if index is not of Datetime type
    if not isinstance(df.index, pd.DatetimeIndex):
        raise TypeError(f"{name}: index must be a DatetimeIndex")

    # Raise an error if dataset is not 8760 rows exactly
    if len(df) != 8760:
        raise ValueError(f"f{name}: expected 8760 rows, got: {len(df)}")

    # Raise an error if dataframe index has duplicates
    if df.index.has_duplicates:
        raise ValueError(f"{name}: index has duplicate timestamps")

    # Raise an error if snapshots do not cover ALL 8760 hours in the year
    if not df.index.equals(expected_index):
        missing = expected_index.difference(df.index)
        extra = df.index.difference(expected_index)

        raise ValueError(
            f"{name}: index does not exactly match hourly {year_to_check}."
            f"Missing examples: {missing[:5].tolist()}. "
            f"Extra examples: {extra[:5].tolist()}. "
        )

    # Raise an error if data contains NaNs
    if df.isna().any().any():
        cols = df.columns[df.isna().any()].tolist()
        raise ValueError(f"{name}: contains NaNs in columns {cols}")

    # Raise an error if data contains at least 1 non-numeric column
    non_numeric_cols = df.select_dtypes(exclude="number").columns.tolist()
    if non_numeric_cols:
        raise TypeError(
            f"All columns have to be numeric.\n"
            f"{name}: non-numeric cols are found {non_numeric_cols}"
        )

## CF range validation
Also add a CF range validator, which will check whether the ranges of values in the CF datasets are between 0.0 - 1.0, is not – raise an error, else print success.

In [594]:
def validate_cf_range(df, name):
    """
    Validate that the CF dataset contains values in the 0.0 = 1.0 range
    """

    # Check whether in the range
    is_in_range = df.ge(0.0) & df.le(1.0)

    # Raise an error if at least 1 row is not in the range
    if not is_in_range.all().all():
        raise ValueError(f"{name}: it is not in the CF (0.0 - 1.0) range.")

## Running validation
Run the full data validation pipeline. Raises an error if something was not passed, else print success.

In [595]:
for name, df in data:
    # Validate all datasets on the main validation pipeline
    validate_df(df, name, 2025, "h")

    # Validate CFs only (in 0.0 - 1.0 range?)
    if name.endswith("cf"):
        validate_cf_range(df, name)

print("Data validation has been successfull!")

Data validation has been successfull!


Since TSAM expects all of the data per regional resolution you're working with (nodal, zonal / whatever geographical area we are working with), in our case – Germany only, so join the features into Germany-specific dataset.

In [596]:
# Join features
data_DE = demand_DE.join([solar_cf_DE,
    onwind_cf_DE,
    ror_cf_DE,
    hydro_inflow_DE])

# Rename a column to demand for clarity
data_DE = data_DE.rename(columns={
    "DE": "demand_DE"})
data_DE.head()

,demand_DE,DE_solar_2025,DE_onwind_2025,DE_ror_2025,DE_hydro_2025
snapshot,,,,,
2025-01-01 00:00:00,28833,0.0,0.716840,0.596077,54
2025-01-01 01:00:00,27781,0.0,0.728613,0.591670,54
2025-01-01 02:00:00,26917,0.0,0.743509,0.584802,54
2025-01-01 03:00:00,27016,0.0,0.760476,0.577704,54
2025-01-01 04:00:00,26870,0.0,0.771249,0.575100,54


# TSAM Clustering
For example, let's use 6 clusters, with period duration of 24 hrs, hierarchical clustering and medoid representation.

API reference: https://tsam.readthedocs.io/en/latest/api/tsam/api/#tsam.api.aggregate

In [597]:
result = tsam.aggregate(
    data=data_DE,
    n_clusters=6, # Number of clusters we want
    period_duration=24, # The period duration in hours (in this case representative day)
    preserve_column_means=True, # Preserve means of the full-resolution features (pre aggregation)
    cluster = tsam.ClusterConfig( # Clustering algorithm configuration
        method="hierarchical",  # Specify how to construct a cluster
        representation="medoid" # Specify what the representative day should be (average or real day)
    )
)

There are a few arguments and options in tsam.aggregate that I would like to point out that may be not so clear from the API documentation, mainly arguments in ClusteringConfig.
- method: there are a lot of methods, but the ones that are the most promising are – medoid, hierarchical, and contiguous.
- representation: same here, lots of options, but the most promising – medoid, and maybe distribution when it matters.
- include_period_sums: might be useful when total day feature shape also matters and not just per-hour shape 

## Looking at the results

API reference: https://tsam.readthedocs.io/en/latest/api/tsam/result/#tsam.result.AggregationResult

What does the output look like?


### Shapes
First, let's look at the shape of the data before & after:

In [598]:
print(f"Num of rows and num of cols in the original full-year dataset: {data_DE.shape}")
print(f"Num of rows and num of cols in the representative output: {result.cluster_representatives.shape}")

Num of rows and num of cols in the original full-year dataset: (8760, 5)
Num of rows and num of cols in the representative output: (144, 5)


The number of features is the same, it was 5 before, and is 5 after, but with 6 clusters and 24 hour period duration (representative day) we reduce the number of hours from 8760 to 144.

Let's also look at the number of original vs representative periods, and timesteps per representative period:


In [599]:
n_rep_periods, n_timesteps_per_period = result.n_clusters, result.n_timesteps_per_period

print(f"Num of original periods: {data_DE.shape[0]}")
print(f"Num of representative periods: {n_rep_periods}")
print(f"Number of timesteps per representative period: {n_timesteps_per_period}")

Num of original periods: 8760
Num of representative periods: 6
Number of timesteps per representative period: 24


### Outputs

#### Cluster representatives

In [600]:
result.cluster_representatives

DE_hydro_2025  DE_onwind_2025  DE_ror_2025  DE_solar_2025  \
  timestep                                                              
0 0            187.109133        0.125963     0.618568            0.0   
  1            187.109133        0.131710     0.615149            0.0   
  2            188.054713        0.137004     0.611909            0.0   
  3            188.054713        0.144796     0.609348            0.0   
  4            189.000293        0.150914     0.607840            0.0   
...                   ...             ...          ...            ...   
5 19            84.040888        0.612087     0.791893            0.0   
  20            84.986468        0.620055     0.773821            0.0   
  21            84.986468        0.603987     0.732510            0.0   
  22            84.986468        0.548563     0.716565            0.0   
  23            84.986468        0.532782     0.715713            0.0   

               demand_DE  
  timestep                
0 0         32410.062815  
  1         31702.972592  
  2         31651.582455  
  3         31458.393606  
  4         31440.311891  
...                  ...  
5 19        38053.461230  
  20        37102.743689  
  21        37389.196121  
  22        35405.917488  
  23        33696.719587  

[144 rows x 5 columns]

In [601]:
result.cluster_representatives.head()

DE_hydro_2025  DE_onwind_2025  DE_ror_2025  DE_solar_2025  \
  timestep                                                              
0 0            187.109133        0.125963     0.618568            0.0   
  1            187.109133        0.131710     0.615149            0.0   
  2            188.054713        0.137004     0.611909            0.0   
  3            188.054713        0.144796     0.609348            0.0   
  4            189.000293        0.150914     0.607840            0.0   

               demand_DE  
  timestep                
0 0         32410.062815  
  1         31702.972592  
  2         31651.582455  
  3         31458.393606  
  4         31440.311891

Index is in the shape of (cluster number, hour of the day), the columns are represented depending on the representation argument we have provided. In this case, it's a real day (medoid), so the values are from a real day.

#### Cluster assigments

In [602]:
result.cluster_assignments

array([5, 1, 1, 1, 4, 5, 5, 1, 1, 1, 1, 2, 2, 1, 1, 1, 1, 1, 2, 2, 1, 1,
       1, 1, 1, 1, 5, 1, 1, 1, 1, 1, 2, 2, 1, 1, 1, 1, 2, 2, 4, 1, 1, 1,
       1, 1, 2, 2, 2, 2, 2, 2, 1, 2, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
       3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 3, 3, 3, 3, 3, 3, 3, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 2,
       2, 2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 2, 2, 2,
       2, 2, 0, 0, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 2, 2, 2, 2, 2,
       0, 0, 2, 2, 2, 2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 5, 5, 2, 2, 2, 2, 4,
       4, 2, 2, 2, 2, 2, 0, 0, 2, 2, 2, 4, 5, 5, 5, 2, 2, 2, 1, 1, 1, 0,
       1, 1, 1, 1, 2, 4, 4, 1, 1, 5, 5, 5, 5, 5, 5,

Let's look at the shape:

In [603]:
result.cluster_assignments.shape[0]

365

It is 365-long array which tells us to which cluster each day belongs to. 

For example: the first day belongs to the cluster #5, second, third, fourth days – to the cluster #1, etc.

#### Cluster assigments

In [604]:
result.assignments

,period_idx,timestep_idx,cluster_idx
snapshot,,,
2025-01-01 00:00:00,0,0,5
2025-01-01 01:00:00,0,1,5
2025-01-01 02:00:00,0,2,5
2025-01-01 03:00:00,0,3,5
2025-01-01 04:00:00,0,4,5
...,...,...,...
2025-12-31 19:00:00,364,19,4
2025-12-31 20:00:00,364,20,4
2025-12-31 21:00:00,364,21,4


Mostly the same data as in the previous cell, but with some additions.

Cluster assigments show us:
- period_idx: which period is it chronologically, e.g. we have set period_duration=24 the aggregation step, so a single period = 24 hours, thus the first day of the year = period #1, second day = period #2, etc.
- timestep_idx: same as period_idx but is on the hour-level inside the period, e.g. with period_duration=24, it ranges from 0-23, if the period duration was 8, then it would range 0-7.
- cluster_idx: the index of a cluster to which the current day is assigned to.

#### Cluster weights

In [605]:
result.cluster_weights

{np.int64(0): np.int64(96),
 np.int64(1): np.int64(78),
 np.int64(2): np.int64(101),
 np.int64(3): np.int64(45),
 np.int64(4): np.int64(29),
 np.int64(5): np.int64(16)}

These are cluster weights, the weight of a cluster being how many days it represents. For example: the first cluster (or representative day) represents 96 days, second 78 days, etc.

#### Reconstructed data

In [606]:
result.reconstructed.head()

,DE_hydro_2025,DE_onwind_2025,DE_ror_2025,DE_solar_2025,demand_DE
snapshot,,,,,
2025-01-01 00:00:00,77.421827,0.580326,0.629560,0.0,32854.492336
2025-01-01 01:00:00,78.367407,0.589765,0.627519,0.0,31891.403096
2025-01-01 02:00:00,78.367407,0.641840,0.624338,0.0,31640.162425
2025-01-01 03:00:00,78.367407,0.720882,0.623942,0.0,31805.752867
2025-01-01 04:00:00,78.367407,0.777625,0.625706,0.0,31439.360221


This is a reconstructed 8760 hour-long dataset with values from each period's representative day (each hour is the data from the same hour of the representative day this snapshot belongs to)

#### Residuals